File: telegram-bot/src/reports.js
/**
 * 보고서 생성 모듈
 * 일일, 주간, 월간 보고서 생성 함수를 제공합니다.
 */

const { formatCurrency, formatPercentage } = require('./utils');

/**
 * 일일 보고서 생성
 * @param {Object} data - 일일 보고서 데이터
 * @returns {string} 마크다운 형식의 보고서
 */
function generateDailyReport(data) {
  let report = `📊 *일일 보고서 (${data.date})*\n\n`;
  
  // 포트폴리오 요약
  report += '*포트폴리오 요약*\n';
  report += `현재 가치: ${formatCurrency(data.portfolioValue)}\n`;
  report += `전일 가치: ${formatCurrency(data.previousValue)}\n`;
  report += `변동: ${formatCurrency(data.change)} (${formatPercentage(data.changePercent)})\n\n`;
  
  // 시장 요약
  report += '*시장 요약*\n';
  report += `한국 시장: ${data.marketSummary.korean}\n`;
  report += `미국 시장: ${data.marketSummary.us}\n\n`;
  
  // 상위 종목
  report += '*상위 종목*\n';
  if (data.topPerformers.length > 0) {
    data.topPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '상위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 하위 종목
  report += '*하위 종목*\n';
  if (data.worstPerformers.length > 0) {
    data.worstPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '하위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 거래 내역
  report += '*오늘의 거래*\n';
  if (data.transactions.length > 0) {
    data.transactions.forEach((tx) => {
      const type = tx.type === 'buy' ? '매수' : '매도';
      report += `${type} ${tx.symbol} ${tx.quantity} 주 @ ${formatCurrency(tx.price)}\n`;
    });
  } else {
    report += '오늘의 거래 내역이 없습니다.\n';
  }
  
  report += '\n';
  
  // AI 분석
  report += '*AI 분석*\n';
  report += data.aiAnalysis || '분석 정보가 없습니다.';
  
  return report;
}

/**
 * 주간 보고서 생성
 * @param {Object} data - 주간 보고서 데이터
 * @returns {string} 마크다운 형식의 보고서
 */
function generateWeeklyReport(data) {
  let report = `📈 *주간 보고서 (${data.startDate} ~ ${data.endDate})*\n\n`;
  
  // 포트폴리오 요약
  report += '*포트폴리오 요약*\n';
  report += `현재 가치: ${formatCurrency(data.portfolioValue)}\n`;
  report += `시작 가치: ${formatCurrency(data.previousValue)}\n`;
  report += `변동: ${formatCurrency(data.change)} (${formatPercentage(data.changePercent)})\n\n`;
  
  // 시장 요약
  report += '*시장 요약*\n';
  report += `한국 시장: ${data.marketSummary.korean}\n`;
  report += `미국 시장: ${data.marketSummary.us}\n\n`;
  
  // 섹터 성과
  report += '*섹터별 성과*\n';
  if (data.sectorPerformance && data.sectorPerformance.labels) {
    data.sectorPerformance.labels.forEach((sector, index) => {
      const value = data.sectorPerformance.values[index];
      report += `${sector}: ${formatPercentage(value)}\n`;
    });
  } else {
    report += '섹터별 성과 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 상위 종목
  report += '*주간 상위 종목*\n';
  if (data.topPerformers.length > 0) {
    data.topPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '상위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 하위 종목
  report += '*주간 하위 종목*\n';
  if (data.worstPerformers.length > 0) {
    data.worstPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '하위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 거래 내역
  report += '*주간 거래 요약*\n';
  if (data.transactions.length > 0) {
    data.transactions.forEach((tx) => {
      const type = tx.type === 'buy' ? '매수' : '매도';
      report += `${tx.date}: ${type} ${tx.symbol} ${tx.quantity} 주 @ ${formatCurrency(tx.price)}\n`;
    });
  } else {
    report += '이번 주 거래 내역이 없습니다.\n';
  }
  
  report += '\n';
  
  // AI 분석
  report += '*AI 주간 분석*\n';
  report += data.aiAnalysis || '분석 정보가 없습니다.';
  
  return report;
}

/**
 * 월간 보고서 생성
 * @param {Object} data - 월간 보고서 데이터
 * @returns {string} 마크다운 형식의 보고서
 */
function generateMonthlyReport(data) {
  let report = `📆 *월간 보고서 (${data.month})*\n\n`;
  
  // 포트폴리오 요약
  report += '*포트폴리오 요약*\n';
  report += `현재 가치: ${formatCurrency(data.portfolioValue)}\n`;
  report += `시작 가치: ${formatCurrency(data.previousValue)}\n`;
  report += `변동: ${formatCurrency(data.change)} (${formatPercentage(data.changePercent)})\n\n`;
  
  // 월별 비교
  report += '*월별 수익률 비교*\n';
  if (data.monthlyComparison && data.monthlyComparison.labels) {
    data.monthlyComparison.labels.forEach((month, index) => {
      const value = data.monthlyComparison.values[index];
      report += `${month}: ${formatPercentage(value)}\n`;
    });
  } else {
    report += '월별 비교 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 시장 요약
  report += '*시장 요약*\n';
  report += `한국 시장: ${data.marketSummary.korean}\n`;
  report += `미국 시장: ${data.marketSummary.us}\n\n`;
  
  // 섹터 성과
  report += '*섹터별 성과*\n';
  if (data.sectorPerformance && data.sectorPerformance.labels) {
    data.sectorPerformance.labels.forEach((sector, index) => {
      const value = data.sectorPerformance.values[index];
      report += `${sector}: ${formatPercentage(value)}\n`;
    });
  } else {
    report += '섹터별 성과 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 상위 종목
  report += '*월간 상위 종목*\n';
  if (data.topPerformers.length > 0) {
    data.topPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '상위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 하위 종목
  report += '*월간 하위 종목*\n';
  if (data.worstPerformers.length > 0) {
    data.worstPerformers.forEach((stock, index) => {
      report += `${index + 1}. ${stock.symbol} (${stock.name}): ${formatPercentage(stock.change)}\n`;
    });
  } else {
    report += '하위 종목 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // 거래 내역 요약
  report += '*월간 거래 요약*\n';
  if (data.transactionSummary) {
    report += `총 거래 횟수: ${data.transactionSummary.totalCount}회\n`;
    report += `매수 횟수: ${data.transactionSummary.buyCount}회\n`;
    report += `매도 횟수: ${data.transactionSummary.sellCount}회\n`;
    report += `총 거래 금액: ${formatCurrency(data.transactionSummary.totalAmount)}\n`;
  } else {
    report += '거래 요약 정보가 없습니다.\n';
  }
  
  report += '\n';
  
  // AI 분석
  report += '*AI 월간 분석 및 전망*\n';
  report += data.aiAnalysis || '분석 정보가 없습니다.';
  
  return report;
}

module.exports = {
  generateDailyReport,
  generateWeeklyReport,
  generateMonthlyReport
};

File: backend/app/routers/reports.py
"""
보고서 관련 API 라우터
"""
import logging
from typing import Dict, List, Any, Optional
from datetime import datetime, timedelta
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

# 로깅 설정
logger = logging.getLogger(__name__)

# 라우터 생성
router = APIRouter(
    prefix="/api/reports",
    tags=["reports"],
    responses={404: {"description": "Not found"}},
)

# 일일 보고서 엔드포인트
@router.get("/daily")
async def get_daily_report(date: Optional[str] = None):
    """
    일일 보고서를 가져옵니다.
    
    Args:
        date: 보고서 날짜 (YYYY-MM-DD, 기본값: 오늘)
    
    Returns:
        일일 보고서 데이터
    """
    try:
        # 날짜 설정
        if not date:
            date = datetime.now().strftime("%Y-%m-%d")
        
        # 샘플 데이터 생성 (실제 구현에서는 데이터베이스에서 가져와야 함)
        report_data = {
            "date": date,
            "portfolioValue": 128.72,
            "previousValue": 127.20,
            "change": 1.52,
            "changePercent": 1.2,
            "marketSummary": {
                "korean": "한국 시장은 최근 반도체와 2차전지 관련주를 중심으로 상승세를 보이고 있습니다.",
                "us": "미국 시장은 기술주를 중심으로 상승세를 유지하고 있습니다."
            },
            "topPerformers": [
                {"symbol": "AAPL", "name": "Apple Inc.", "change": 2.1},
                {"symbol": "NVDA", "name": "NVIDIA Corp.", "change": 3.2},
                {"symbol": "005930", "name": "삼성전자", "change": 1.5}
            ],
            "worstPerformers": [
                {"symbol": "TSLA", "name": "Tesla Inc.", "change": -1.5},
                {"symbol": "AMZN", "name": "Amazon.com Inc.", "change": -0.8},
                {"symbol": "035720", "name": "카카오", "change": -1.2}
            ],
            "transactions": [
                {"type": "buy", "symbol": "AAPL", "quantity": 0.1, "price": 182.30}
            ],
            "aiAnalysis": "시장은 전반적으로 긍정적인 흐름을 보이고 있으며, 특히 기술 섹터가 강세를 보이고 있습니다. 단기적으로는 중앙은행의 금리 정책과 인플레이션 지표를 주시할 필요가 있습니다."
        }
        
        return report_data
    
    except Exception as e:
        logger.error(f"일일 보고서 가져오기 중 오류 발생: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"일일 보고서 가져오기 중 오류 발생: {str(e)}"
        )

# 주간 보고서 엔드포인트
@router.get("/weekly")
async def get_weekly_report(end_date: Optional[str] = None):
    """
    주간 보고서를 가져옵니다.
    
    Args:
        end_date: 보고서 종료 날짜 (YYYY-MM-DD, 기본값: 오늘)
    
    Returns:
        주간 보고서 데이터
    """
    try:
        # 날짜 설정
        if not end_date:
            end_date = datetime.now().strftime("%Y-%m-%d")
        
        # 시작 날짜 계산 (종료 날짜로부터 7일 전)
        end_date_obj = datetime.strptime(end_date, "%Y-%m-%d")
        start_date_obj = end_date_obj - timedelta(days=7)
        start_date = start_date_obj.strftime("%Y-%m-%d")
        
        # 샘플 데이터 생성 (실제 구현에서는 데이터베이스에서 가져와야 함)
        report_data = {
            "startDate": start_date,
            "endDate": end_date,
            "portfolioValue": 128.72,
            "previousValue": 125.50,
            "change": 3.22,
            "changePercent": 2.57,
            "marketSummary": {
                "korean": "한국 시장은 최근 반도체와 2차전지 관련주를 중심으로 상승세를 보이고 있습니다.",
                "us": "미국 시장은 기술주를 중심으로 상승세를 유지하고 있습니다."
            },
            "sectorPerformance": {
                "labels": ["기술", "금융", "헬스케어", "소비재", "에너지"],
                "values": [3.2, 1.5, 2.1, -0.8, -1.2]
            },
            "topPerformers": [
                {"symbol": "AAPL", "name": "Apple Inc.", "change": 4.5},
                {"symbol": "NVDA", "name": "NVIDIA Corp.", "change": 6.8},
                {"symbol": "005930", "name": "삼성전자", "change": 3.2}
            ],
            "worstPerformers": [
                {"symbol": "TSLA", "name": "Tesla Inc.", "change": -3.2},
                {"symbol": "AMZN", "name": "Amazon.com Inc.", "change": -1.5},
                {"symbol": "035720", "name": "카카오", "change": -2.5}
            ],
            "transactions": [
                {"date": start_date, "type": "buy", "symbol": "AAPL", "quantity": 0.1, "price": 175.30},
                {"date": end_date, "type": "buy", "symbol": "NVDA", "quantity": 0.05, "price": 850.20}
            ],
            "aiAnalysis": "이번 주 시장은 전반적으로 긍정적인 흐름을 보였으며, 특히 기술 섹터가 강세를 보였습니다. 다음 주에는 주요 기업들의 실적 발표가 예정되어 있어 변동성이 커질 수 있습니다."
        }
        
        return report_data
    
    except Exception as e:
        logger.error(f"주간 보고서 가져오기 중 오류 발생: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"주간 보고서 가져오기 중 오류 발생: {str(e)}"
        )

# 월간 보고서 엔드포인트
@router.get("/monthly")
async def get_monthly_report(month: Optional[str] = None):
    """
    월간 보고서를 가져옵니다.
    
    Args:
        month: 보고서 월 (YYYY-MM, 기본값: 현재 월)
    
    Returns:
        월간 보고서 데이터
    """
    try:
        # 날짜 설정
        if not month:
            month = datetime.now().strftime("%Y-%m")
        
        # 샘플 데이터 생성 (실제 구현에서는 데이터베이스에서 가져와야 함)
        report_data = {
            "month": month,
            "portfolioValue": 128.72,
            "previousValue": 120.30,
            "change": 8.42,
            "changePercent": 7.0,
            "monthlyComparison": {
                "labels": ["1주차", "2주차", "3주차", "4주차"],
                "values": [122.50, 124.80, 126.30, 128.72]
            },
            "marketSummary": {
                "korean": "한국 시장은 최근 반도체와 2차전지 관련주를 중심으로 상승세를 보이고 있습니다.",
                "us": "미국 시장은 기술주를 중심으로 상승세를 유지하고 있습니다."
            },
            "sectorPerformance": {
                "labels": ["기술", "금융", "헬스케어", "소비재", "에너지"],
                "values": [8.5, 3.2, 5.1, -1.5, -2.8]
            },
            "topPerformers": [
                {"symbol": "AAPL", "name": "Apple Inc.", "change": 10.2},
                {"symbol": "NVDA", "name": "NVIDIA Corp.", "change": 15.5},
                {"symbol": "005930", "name": "삼성전자", "change": 7.8}
            ],
            "worstPerformers": [
                {"symbol": "TSLA", "name": "Tesla Inc.", "change": -5.5},
                {"symbol": "AMZN", "name": "Amazon.com Inc.", "change": -3.2},
                {"symbol": "035720", "name": "카카오", "change": -4.5}
            ],
            "transactionSummary": {
                "totalCount": 5,
                "buyCount": 3,
                "sellCount": 2,
                "totalAmount": 150.25
            },
            "aiAnalysis": "이번 달 시장은 전반적으로 긍정적인 흐름을 보였으며, 특히 기술 섹터가 강세를 보였습니다. 다음 달에는 중앙은행의 금리 결정과 주요 경제 지표 발표가 예정되어 있어 이에 대한 주의가 필요합니다."
        }
        
        return report_data
    
    except Exception as e:
        logger.error(f"월간 보고서 가져오기 중 오류 발생: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"월간 보고서 가져오기 중 오류 발생: {str(e)}"
        )
